packages/react/src/util.ts

```diff
@@ -0,0 +1,27 @@
+import { PenumbraInjection, PenumbraSymbol } from '@penumbra-zone/client';
+
+export const keyOfInjection = (injection?: PenumbraInjection) =>
+  Object.entries(window[PenumbraSymbol] ?? {}).find(
+    ([keyOrigin, valueInjection]) =>
+      keyOrigin &&
+      // matching injection
+      valueInjection === injection,
+  )?.[0];
+
+export const injectionOfKey = (keyOrigin?: string) =>
+  keyOrigin ? window[PenumbraSymbol]?.[keyOrigin] : undefined;
+
+export const assertStringIsOrigin = (s?: string) => {
+  if (!s || new URL(s).origin !== s) {
```

Comment:
technically, a provider could identify itself as any origin hosting a manifest. there is nothing, technically, preventing a chrome extension or even another embedded script in the dapp itself from creating a provider record that uses any protocol and any transport.

in the interest of versatility and forward compatibility, especially with other browsers that may implement provider extensions, i haven't restricted this.

there's no security downside, because anything capable of injecting a provider has already pwned the page.